sway-core/src/semantic_analysis/ast_node/declaration/declaration.rs

```diff
@@ -209,20 +212,84 @@ impl ty::TyDecl {
                     warnings,
                     errors
                 );
-                check!(
-                    ctx.namespace.insert_trait_implementation(
-                        impl_trait.trait_name.clone(),
-                        impl_trait.trait_type_arguments.clone(),
-                        impl_trait.implementing_for.type_id,
-                        &impl_trait.items,
-                        &impl_trait.span,
-                        true,
-                        engines,
-                    ),
-                    return err(warnings, errors),
-                    warnings,
-                    errors
-                );
+
+                // Set of all type IDs that have corresponding items implemented for them in this
+                // `impl` block. In most cases, this is a single type ID that is equal to
+                // `impl_trait.implementing_for.type_id`.
+                let mut processed_ids = HashSet::new();
+
+                // This is a map from type IDs to a vector of impl items. In most cases, all the
+                // items are available to `impl_trait.implementing_for.type_id` (i.e. `Self`).
+                // However, there are situations where `self`, in some of the impl methods, has a
+                // type ascription that is different from `Self`. In that case, we have to keep
+                // track of which impl items are available to which type ID.
```

Comment:
I was somewhat under the impression that this does not work in Sway just yet 🤔 but, to my surprise, it does seem to work fine without any issues. As you said, we can introduce rules around allowing `impl` for upstream generic types as long as their generic input type is filled with a type from the current package. It might just be okay to deviate from Rust in that regard.

Let me nuke all the code related to `self` type ascriptions, or at least shelve it in case we need it in the future.